oak/server/oak_runtime.cc

```diff
@@ -107,7 +107,7 @@ std::string OakRuntime::NextNodeName(const std::string& config_name,
 // Create (but don't start) a new Node instance.  Return a borrowed pointer to
 // the new Node (or nullptr on failure).
 OakNode* OakRuntime::CreateNode(const std::string& config_name, const std::string& entrypoint_name,
-                                std::string* node_name) {
+                                const std::string& creator_name, std::string* node_name) {
```

Comment:
I'm not sure I see the value of this `creator_name` param, ideally `node_name` should be sufficient to figure out the identity of the node. If the creator wanted to make it explicit, it could add its own name as part of the new node name, I guess (though I don't think we should actually do that in code). Were you trying to address a specific issue with this?

Reply:
The main problem I was trying to solve is that for Applications that have multiple copies of the same Node (e.g. the [`BackendRoundtrip`](https://github.com/project-oak/oak/blob/master/examples/abitest/module_0/rust/src/lib.rs#L1022-L1097) testcase), the logging is currently too fungible to be useful:

```log
2020-03-02 15:46:01  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 218890223900607730 } }, out=Handle { id: 12260816694213148413 }
2020-03-02 15:46:01  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 8320307275176590207 } }, out=Handle { id: 10297784421300438978 }
2020-03-02 15:46:01  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 8329056563568872758 } }, out=Handle { id: 12685423956684209417 }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 2652641776716562875 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 7725204717666180153 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 2950376055256300427 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[1]=ReadHandle { handle: Handle { id: 2652641776716562875 } }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 5472346579660996369 } }: {"msg":"aaaxxx"}
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 14635611646120412975 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 6072819210769413298 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 11499504671846902189 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 9007958221342433685 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 741350813642357911 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 12536364673992575351 } to waiting set
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[2]=ReadHandle { handle: Handle { id: 7725204717666180153 } }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 8589544076168364742 } }: {"msg":"aaaxxx"}
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[3]=ReadHandle { handle: Handle { id: 2950376055256300427 } }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:46:04  INFO  logging_node.cc : 50 : LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 18184210530674926808 } }: {"msg":"aaaxxx"}
```

This change allows the creator of each logging Node to be remembered and logged, yielding logs that are more helpful &ndash; it's possible to figure out which Node got which message:

```log
2020-03-02 15:49:57  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 15581786514528961327 } }, out=Handle { id: 12886977840758777362 }
2020-03-02 15:49:57  INFO  logging_node.cc : 50 : {backend-config-1-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 629742161666882353 } }, out=Handle { id: 17392952505014550743 }
2020-03-02 15:49:57  INFO  logging_node.cc : 50 : {backend-config-2-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 12412937325566452008 } }, out=Handle { id: 11792555408562386829 }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 17327170281632557498 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 12403852401395599495 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 4596109841017128518 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[1]=ReadHandle { handle: Handle { id: 17327170281632557498 } }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 15441846858513127221 } }: {"msg":"aaaxxx"}
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-1-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 5547489475360994078 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-1-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 5919197871457181556 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-1-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 6691320800906629762 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-2-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 16622342038908165713 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-2-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 3834253999603801904 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-2-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 11125563812863221933 } to waiting set
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[2]=ReadHandle { handle: Handle { id: 12403852401395599495 } }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 6210276338543828832 } }: {"msg":"aaaxxx"}
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[3]=ReadHandle { handle: Handle { id: 4596109841017128518 } }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:50:02  INFO  logging_node.cc : 50 : {backend-config-0-backend_oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 16941296889394752364 } }: {"msg":"aaaxxx"}
```

The prefix there is the (entirely internal) Node name of who created the logging Node, and so shows the config name and entrypoint name of the Node instance making the log (assuming there have been no handle-passing shenanigans going on).

I guess we could alternatively just use the Node name of the logging Node itself, which would be unique and therefore distinguishable:

```log
2020-03-02 15:48:24  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 12213789355667294910 } }, out=Handle { id: 3314670794225268986 }
2020-03-02 15:48:24  INFO  logging_node.cc : 50 : {logging-config-2-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 2566390685214579613 } }, out=Handle { id: 7742444656904688837 }
2020-03-02 15:48:24  INFO  logging_node.cc : 50 : {logging-config-3-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 58 : backend node: in=ReadHandle { handle: Handle { id: 1458673124996920667 } }, out=Handle { id: 5840554695612818549 }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 17806286846115458112 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 11692316322676417325 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 16564610347881176357 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[1]=ReadHandle { handle: Handle { id: 17806286846115458112 } }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 4921531041891495006 } }: {"msg":"aaaxxx"}
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-2-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 9982898099008659992 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-2-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 12423129559091251713 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-2-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 17139404759693640201 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-3-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 12204958912891550573 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-3-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 12638919110422873748 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-3-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 72 : add new handle Handle { id: 6868809729784251879 } to waiting set
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[2]=ReadHandle { handle: Handle { id: 11692316322676417325 } }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 9402025034155290889 } }: {"msg":"aaaxxx"}
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 83 : message available on waiting handle[3]=ReadHandle { handle: Handle { id: 16564610347881176357 } }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 97 : received frontend request: InternalMessage { msg: "aaa" }
2020-03-02 15:48:26  INFO  logging_node.cc : 50 : {logging-config-1-oak_main} LOG: INFO  examples/abitest/module_1/rust/src/lib.rs : 105 : send serialized message to new channel WriteHandle { handle: Handle { id: 3676358255456401078 } }: {"msg":"aaaxxx"}
```

but that would have the downside that any logs from library code that's shared between different Node types would have to be manually tracked back to the logging Node creation event for correlation.

Reply:
(Thinking on this more, including the logging node name is probably good enough, so I've shifted to that.)